File: android/app/src/main/java/app/notomorrow/net/MockBackendClient.kt
package app.notomorrow.net

import app.notomorrow.R
import app.notomorrow.model.AttendanceStatus
import app.notomorrow.model.HeadsUpKind
import app.notomorrow.model.MealSlot
import app.notomorrow.model.Participant
import app.notomorrow.net.dto.AIEstimate
import app.notomorrow.net.dto.AIFood
import app.notomorrow.net.dto.AttendanceDto
import app.notomorrow.net.dto.HeadsUpDto
import app.notomorrow.net.dto.Me
import app.notomorrow.net.dto.Partner
import app.notomorrow.net.dto.PartnerState
import app.notomorrow.net.dto.ScheduleDto
import app.notomorrow.net.dto.Session
import io.ktor.http.Url
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import java.time.Instant
import java.time.LocalDate
import java.util.UUID

/**
 * In-memory backend for offline demos — the port of
 * `NoTomorrow/Services/MockBackendClient.swift`.
 *
 * Partner "Tomek" shares your schedule, confirms **20 s** after you do, and
 * answers a "can't make it" after **10 s**. Pairing survives relaunch (through
 * [MockPairingStore]) so the Bro tab stays coherent. A [Mutex] plays the part of
 * Swift's `actor` isolation.
 */
class MockBackendClient(
    private val strings: MockStrings,
    private val pairing: MockPairingStore = InMemoryMockPairingStore(),
    override val baseUrl: Url = Url("https://mock.notomorrow.app"),
    private val scope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Default),
) : BackendClient {

    private val mutex = Mutex()

    private var session: Session? = null
    private var schedule: ScheduleDto = ScheduleDto.of(listOf(1, 3, 5), 18 * 60)
    private var partner: Partner? = null

    /** Explicit partner attendance, keyed by day. Anything not here is derived from the schedule. */
    private val partnerOverrides = mutableMapOf<LocalDate, AttendanceDto>()
    private val myAttendance = mutableMapOf<LocalDate, AttendanceDto>()
    private val headsUps = mutableListOf<HeadsUpDto>()
    private var replyJob: Job? = null
    private var confirmJob: Job? = null

    init {
        if (pairing.isPaired()) {
            partner = Partner(PARTNER_ID, PARTNER_NAME, Instant.now().minusSeconds(21L * 86_400))
        }
    }

    // MARK: Auth

    override suspend fun signInApple(identityToken: String, authorizationCode: String): Session =
        mutex.withLock { makeSession("mock-apple") }

    override suspend fun signInGoogle(idToken: String): Session =
        mutex.withLock { makeSession("mock-google") }

    override suspend fun signIn(username: String, password: String): Session =
        mutex.withLock { makeSession("mock-$username") }

    override suspend fun register(username: String, password: String, email: String?): Session =
        mutex.withLock { makeSession("mock-$username") }

    private fun makeSession(userId: String): Session =
        Session(
            accessToken = "mock-access-${UUID.randomUUID()}",
            refreshToken = "mock-refresh-${UUID.randomUUID()}",
            userId = userId,
        ).also { session = it }

    override suspend fun logout(refreshToken: String) = mutex.withLock { session = null }

    override suspend fun me(): Me = mutex.withLock {
        Me(
            id = session?.userId ?: "mock-user",
            username = MOCK_USERNAME,
            displayName = MOCK_DISPLAY_NAME,
            email = null,
            partner = partner,
            pairCode = PAIR_CODE,
        )
    }

    override suspend fun updateMe(locale: String?, timeZone: String?) = Unit

    // MARK: Pairing

    override suspend fun createPairCode(): String = PAIR_CODE

    override suspend fun pair(code: String): Partner = mutex.withLock {
        val trimmed = code.trim()
        if (trimmed.length != PAIR_CODE_LENGTH) throw BackendError.Server("Invalid code")
        Partner(PARTNER_ID, PARTNER_NAME, Instant.now()).also { setPartner(it) }
    }

    override suspend fun unpair() = mutex.withLock { unpairLocked() }

    private fun unpairLocked() {
        setPartner(null)
        partnerOverrides.clear()
        headsUps.clear()
        confirmJob?.cancel()
        replyJob?.cancel()
    }

    private fun setPartner(value: Partner?) {
        partner = value
        pairing.setPaired(value != null)
    }

    override suspend fun pushSchedule(schedule: ScheduleDto) = mutex.withLock {
        this.schedule = schedule.sorted()
    }

    override suspend fun partnerState(): PartnerState = mutex.withLock {
        val current = partner ?: throw BackendError.Server("Not paired")
        PartnerState(
            partnerName = current.name,
            partnerSchedule = schedule,
            attendance = partnerAttendance() + myAttendance.values.sortedBy { it.day },
            headsUps = headsUps.toList(),
        )
    }

    /**
     * Tomek's record: every gym day in the last three weeks attended (one missed
     * with a note so the log has something to show), today planned until he
     * confirms, plus any explicit overrides.
     */
    private fun partnerAttendance(): List<AttendanceDto> {
        val today = LocalDate.now()
        val rows = mutableListOf<AttendanceDto>()
        val pastGymDays = (1..21)
            .map { today.minusDays(it.toLong()) }
            .filter { schedule.isGymDay(it.dayOfWeek.value) }
        pastGymDays.forEachIndexed { index, day ->
            val override = partnerOverrides[day]
            if (override != null) {
                rows.add(override)
                return@forEachIndexed
            }
            // The third-most-recent gym day is a miss; the last two are attended.
            rows.add(
                if (index == 2) {
                    AttendanceDto(
                        day = day,
                        participant = Participant.Partner,
                        status = AttendanceStatus.Missed,
                        reason = "sick",
                        note = strings.string(R.string.mock_partner_sickNote),
                    )
                } else {
                    AttendanceDto(day, Participant.Partner, AttendanceStatus.Attended)
                }
            )
        }
        if (schedule.isGymDay(today.dayOfWeek.value)) {
            rows.add(
                partnerOverrides[today]
                    ?: AttendanceDto(today, Participant.Partner, AttendanceStatus.Planned)
            )
        }
        return rows.sortedBy { it.day }
    }

    override suspend fun setAttendance(
        day: LocalDate,
        status: AttendanceStatus,
        reason: String?,
        note: String?,
        makeUpDay: LocalDate?,
    ) {
        mutex.withLock {
            myAttendance[day] = AttendanceDto(day, Participant.Me, status, reason, note, makeUpDay)
            if (partner == null || status != AttendanceStatus.Confirmed || day != LocalDate.now()) return
            confirmJob?.cancel()
            confirmJob = scope.launch {
                delay(PARTNER_CONFIRM_DELAY_MS)
                mutex.withLock { partnerConfirms(day) }
            }
        }
    }

    private fun partnerConfirms(day: LocalDate) {
        if (partner == null) return
        partnerOverrides[day] = AttendanceDto(day, Participant.Partner, AttendanceStatus.Confirmed)
    }

    override suspend fun sendHeadsUp(kind: HeadsUpKind, text: String, sessionDay: LocalDate) {
        mutex.withLock {
            headsUps.add(
                HeadsUpDto(
                    id = UUID.randomUUID().toString(),
                    fromMe = true,
                    kind = kind,
                    text = text,
                    sessionDay = sessionDay,
                    sentAt = Instant.now(),
                )
            )
            if (partner == null || kind != HeadsUpKind.CantMakeIt) return
            replyJob?.cancel()
            replyJob = scope.launch {
                delay(PARTNER_REPLY_DELAY_MS)
                mutex.withLock { partnerReplies(sessionDay) }
            }
        }
    }

    private fun partnerReplies(day: LocalDate) {
        if (partner == null) return
        headsUps.add(
            HeadsUpDto(
                id = UUID.randomUUID().toString(),
                fromMe = false,
                kind = HeadsUpKind.Custom,
                text = strings.string(R.string.mock_partner_reply),
                sessionDay = day,
                sentAt = Instant.now(),
            )
        )
    }

    override suspend fun registerPushToken(token: String, platform: String) = Unit

    override suspend fun unregisterPushToken(token: String) = Unit

    // MARK: AI

    override suspend fun estimate(
        imageJpeg: ByteArray,
        meal: MealSlot,
        locale: String,
        anthropicKey: String?,
        notes: String,
    ): AIEstimate {
        delay(ESTIMATE_DELAY_MS)
        return AIEstimate(
            foods = listOf(
                AIFood.of(strings.string(R.string.mock_food_chicken), 180.0, 297.0, 56.0, 0.0, 6.5, 0.92),
                AIFood.of(strings.string(R.string.mock_food_rice), 220.0, 286.0, 5.9, 62.0, 0.6, 0.84),
                AIFood.of(strings.string(R.string.mock_food_broccoli), 90.0, 31.0, 2.5, 6.0, 0.4, 0.88),
                AIFood.of(strings.string(R.string.mock_food_oliveOil), 14.0, 124.0, 0.0, 0.0, 14.0, 0.35, isGuess = true),
            ),
            overallConfidence = 0.78,
        )
    }

    override suspend fun deleteAccount() = mutex.withLock {
        session = null
        unpairLocked()
        myAttendance.clear()
    }

    companion object {
        const val PARTNER_NAME: String = "Tomek"
        const val PARTNER_ID: String = "mock-tomek"
        const val PAIR_CODE: String = "NT-7K4Q"

        /** `BroService.normalizedCode` always produces `NT-XXXX`; any 7-char code pairs. */
        const val PAIR_CODE_LENGTH: Int = 7

        private const val MOCK_USERNAME = "kuba"
        private const val MOCK_DISPLAY_NAME = "Kuba"

        const val PARTNER_CONFIRM_DELAY_MS: Long = 20_000
        const val PARTNER_REPLY_DELAY_MS: Long = 10_000
        const val ESTIMATE_DELAY_MS: Long = 1_200
    }
}

File: NoTomorrow/Services/MockBackendClient.swift
import Foundation

/// In-memory backend for offline demos. Partner "Tomek" shares your schedule, confirms 20 s after you do,
/// and answers a "can't make it" after 10 s. Pairing survives relaunch (UserDefaults) so the Bro tab stays coherent.
actor MockBackendClient: BackendClient {
    nonisolated let baseURL: URL

    static let partnerName = "Tomek"
    static let partnerId = "mock-tomek"
    static let pairCode = "NT-7K4Q"

    private var session: Session?
    private var schedule = ScheduleDTO(weekdays: [1, 3, 5], defaultMinuteOfDay: 18 * 60)
    private var partner: Partner? {
        didSet { UserDefaults.standard.set(partner != nil, forKey: Keys.paired) }
    }
    /// Explicit partner attendance, keyed by start of day. Anything not here is derived from the schedule.
    private var partnerOverrides: [Date: AttendanceDTO] = [:]
    private var myAttendance: [Date: AttendanceDTO] = [:]
    private var headsUps: [HeadsUpDTO] = []
    private var replyTask: Task<Void, Never>?
    private var confirmTask: Task<Void, Never>?

    private enum Keys { static let paired = "nt.mock.paired" }

    init(baseURL: URL = URL(string: "https://mock.notomorrow.app")!) {
        self.baseURL = baseURL
        if UserDefaults.standard.bool(forKey: Keys.paired) {
            partner = Partner(id: Self.partnerId, name: Self.partnerName, pairedAt: .now.addingTimeInterval(-21 * 86_400))
        }
    }

    // MARK: Auth

    func signIn(apple identityToken: String, authorizationCode: String) async throws -> Session { makeSession(userId: "mock-apple") }
    func signIn(google idToken: String) async throws -> Session { makeSession(userId: "mock-google") }
    func signIn(username: String, password: String) async throws -> Session { makeSession(userId: "mock-\(username)") }
    func register(username: String, password: String, email: String?) async throws -> Session { makeSession(userId: "mock-\(username)") }

    private func makeSession(userId: String) -> Session {
        let s = Session(accessToken: "mock-access-\(UUID().uuidString)", refreshToken: "mock-refresh-\(UUID().uuidString)", userId: userId)
        session = s
        return s
    }

    func logout(refreshToken: String) async throws { session = nil }

    func me() async throws -> Me {
        Me(id: session?.userId ?? "mock-user", username: "kuba", displayName: "Kuba", email: nil, partner: partner, pairCode: Self.pairCode)
    }

    func updateMe(locale: String?, timeZone: String?) async throws {}

    // MARK: Pairing

    func createPairCode() async throws -> String { Self.pairCode }

    func pair(withCode code: String) async throws -> Partner {
        let trimmed = code.trimmingCharacters(in: .whitespacesAndNewlines)
        guard trimmed.count == 7 else { throw BackendError.server("Invalid code") }
        let p = Partner(id: Self.partnerId, name: Self.partnerName, pairedAt: .now)
        partner = p
        return p
    }

    func unpair() async throws {
        partner = nil
        partnerOverrides = [:]
        headsUps = []
        confirmTask?.cancel()
        replyTask?.cancel()
    }

    func pushSchedule(_ s: ScheduleDTO) async throws { schedule = s }

    func partnerState() async throws -> PartnerState {
        guard let partner else { throw BackendError.server("Not paired") }
        return PartnerState(partnerName: partner.name, partnerSchedule: schedule,
                            attendance: partnerAttendance() + myAttendance.values.sorted { $0.day < $1.day },
                            headsUps: headsUps)
    }

    /// Tomek's record: every gym day in the last three weeks attended (one missed with a note for the log),
    /// today planned until he confirms, plus any explicit overrides.
    private func partnerAttendance() -> [AttendanceDTO] {
        let cal = Calendar.current
        let today = cal.startOfDay(for: .now)
        var rows: [AttendanceDTO] = []
        var pastGymDays: [Date] = []
        for offset in stride(from: 1, through: 21, by: 1) {
            guard let day = cal.date(byAdding: .day, value: -offset, to: today) else { continue }
            if schedule.isGymDay(cal.isoWeekday(for: day)) { pastGymDays.append(day) }
        }
        for (index, day) in pastGymDays.enumerated() {
            if let override = partnerOverrides[day] { rows.append(override); continue }
            // The third-most-recent gym day is a miss so the log has something to show; the last two are attended.
            if index == 2 {
                rows.append(AttendanceDTO(day: day, participant: .partner, status: .missed, reason: "sick",
                                          note: String(localized: "mock.partner.sickNote"), makeUpDay: nil))
            } else {
                rows.append(AttendanceDTO(day: day, participant: .partner, status: .attended, reason: nil, note: nil, makeUpDay: nil))
            }
        }
        if schedule.isGymDay(cal.isoWeekday(for: today)) {
            rows.append(partnerOverrides[today]
                        ?? AttendanceDTO(day: today, participant: .partner, status: .planned, reason: nil, note: nil, makeUpDay: nil))
        }
        return rows.sorted { $0.day < $1.day }
    }

    func setAttendance(day: Date, status: AttendanceStatus, reason: String?, note: String?, makeUpDay: Date?) async throws {
        let day = Calendar.current.startOfDay(for: day)
        myAttendance[day] = AttendanceDTO(day: day, participant: .me, status: status, reason: reason, note: note, makeUpDay: makeUpDay)
        guard partner != nil, status == .confirmed, Calendar.current.isDateInToday(day) else { return }
        confirmTask?.cancel()
        confirmTask = Task {
            try? await Task.sleep(for: .seconds(20))
            guard !Task.isCancelled else { return }
            partnerConfirms(day)
        }
    }

    private func partnerConfirms(_ day: Date) {
        guard partner != nil else { return }
        partnerOverrides[day] = AttendanceDTO(day: day, participant: .partner, status: .confirmed, reason: nil, note: nil, makeUpDay: nil)
    }

    func sendHeadsUp(kind: HeadsUpKind, text: String, sessionDay: Date) async throws {
        let day = Calendar.current.startOfDay(for: sessionDay)
        headsUps.append(HeadsUpDTO(id: UUID().uuidString, fromMe: true, kind: kind, text: text, sessionDay: day, sentAt: .now))
        guard partner != nil, kind == .cantMakeIt else { return }
        replyTask?.cancel()
        replyTask = Task {
            try? await Task.sleep(for: .seconds(10))
            guard !Task.isCancelled else { return }
            partnerReplies(day)
        }
    }

    private func partnerReplies(_ day: Date) {
        guard partner != nil else { return }
        headsUps.append(HeadsUpDTO(id: UUID().uuidString, fromMe: false, kind: .custom,
                                   text: String(localized: "mock.partner.reply"), sessionDay: day, sentAt: .now))
    }

    func registerPushToken(_ token: Data) async throws {}

    // MARK: AI

    func estimate(imageJPEG: Data, meal: MealSlot, locale: String, anthropicKey: String?, notes: String) async throws -> AIEstimate {
        try await Task.sleep(for: .seconds(1.2))
        return AIEstimate(foods: [
            AIFood(name: String(localized: "mock.food.chicken"), grams: 180, kcal: 297, protein: 56, carbs: 0, fat: 6.5, confidence: 0.92),
            AIFood(name: String(localized: "mock.food.rice"), grams: 220, kcal: 286, protein: 5.9, carbs: 62, fat: 0.6, confidence: 0.84),
            AIFood(name: String(localized: "mock.food.broccoli"), grams: 90, kcal: 31, protein: 2.5, carbs: 6, fat: 0.4, confidence: 0.88),
            AIFood(name: String(localized: "mock.food.oliveOil"), grams: 14, kcal: 124, protein: 0, carbs: 0, fat: 14, confidence: 0.35, isGuess: true),
        ], overallConfidence: 0.78)
    }

    func deleteAccount() async throws {
        session = nil
        try await unpair()
        myAttendance = [:]
    }
}
